### PR TITLE
Add option to provide transaction signature keys to cleos

### PR DESCRIPTION
**Pull request by brianjohnson5972**
## Change Description
Added "--sign-with <public key>" or "--sign-with [ <public key 1>, <public key 2>, ... ]" option to cleos subcommands that result in one signed transaction being sent to nodeos, skipping the step of requesting nodeos for the keys required for signing the transaction.
#8199

## Consensus Changes
- [ ] Consensus Changes

## API Changes
- [x] API Changes

The following cleos subcommands will now allow signing with one key ("--sign-with <public key>") or multiple keys ( "--sign-with [ <public key 1>, <public key 2>, ... ]"):
"set action permission"
"system regproducer"
"system unregproducer"
"vote producer proxy"
"vote producer prods"
"vote producer approve"
"vote producer unapprove"
"system delegatebw"
"system undelegatebw"
"system bidname"
"system buyram"
"system sellram"
"system claimrewards"
"system regproxy"
"system unregproxy"
"system canceldelay"
"system rex deposit"
"system rex withdraw"
"system rex buyrex"
"system rex lendrex"
"system rex unstaketorex"
"system rex sellrex"
"system rex cancelrexorder"
"system rex mvtosavings"
"system rex mvfromsavings"
"system rex rentcpu"
"system rex rentnet"
"system rex fundcpuloan"
"system rex fundnetloan"
"system rex defcpuloan"
"system rex defnetloan"
"system rex consolidate"
"system rex updaterex"
"system rex rexexec"
"system rex closerex"
"system rex "
"set contract"
"set code"
"set abi"
"transfer"
"push action"
"push transaction"
"multisig propose"
"multisig propose_trx"
"multisig approve"
"multisig unapprove"
"multisig invalidate"
"multisig cancel"
"multisig exec"
"wrap exec"

## Documentation Additions
- [x] Documentation Additions

### CODE REVIEW

File: programs/cleos/main.cpp [hunk 1]
```diff
@@ -317,7 +357,8 @@ fc::variant push_transaction( signed_transaction& trx, packed_transaction::compr
             ref_block = call(get_block_func, fc::mutable_variant_object("block_num_or_id", tx_ref_block_num_or_id));
             ref_block_id = ref_block["id"].as<block_id_type>();
          }
-      } EOS_RETHROW_EXCEPTIONS(invalid_ref_block_exception, "Invalid reference block num or id: ${block_num_or_id}", ("block_num_or_id", tx_ref_block_num_or_id));
+      } EOS_RETHROW_EXCEPTIONS(invalid_ref_block_exception, "Invalid refere"
+                                                            "nce block num or id: ${block_num_or_id}", ("block_num_or_id", tx_ref_block_num_or_id));
```
Comment (heifner) [inline programs/cleos/main.cpp]: Was your editor being "helpful"? Lets not break a line in the middle of a word.

File: programs/cleos/main.cpp [hunk 2]
```diff
@@ -221,6 +221,45 @@ void add_standard_transaction_options(CLI::App* cmd, string default_permission =
...
+public:
+   signing_keys_option() {}
+   void add_option(CLI::App* cmd) {
+      cmd->add_option("--sign-with", public_key_json, localized("The public key or json array of public keys to sign with"));
+   }
+
+   std::vector<public_key_type> get_keys() {
+      std::vector<public_key_type> signing_keys;
+      if (!public_key_json.empty()) {
+         if (boost::istarts_with(public_key_json, "EOS") || boost::istarts_with(public_key_json, "PUB_R1")) {
```
Comment (heifner) [inline programs/cleos/main.cpp]: How about adding a static method `is_public_key_str()` that checks for "EOS' and "PUB_R1". Use it here and in `parse_json_authority_or_key`.

Also, shouldn't the valid strings be "EOS", "PUB_R1", "PUB_K1", and "PUB_WA" ? @spoonincode

File: programs/cleos/main.cpp (hunk 1 above)
Comment (brianjohnson5972) [reply programs/cleos/main.cpp]: I guess, I'll change it back

File: programs/cleos/main.cpp (hunk 2 above)
Comment (spoonincode) [reply programs/cleos/main.cpp]: Yes those are the 4 current valid prefixes. Though, keosd can't sign PUB_WA so I'm not sure the usefulness of it being handled; probably still want it for completeness sake

**Pull request merged**